info/src/main/java/edu/artic/info/AccessMemberCardViewModel.kt

```diff
@@ -0,0 +1,225 @@
+
+/**
+ * Validates the user provided member information.
+ * Switches [AccessMemberCardFragment] view state based on [DisplayMode].
+ *
+ * When [DisplayMode]
+ * * is [DisplayMode.DisplayForm] form is displayed
+ * * is [DisplayMode.UpdateForm] the filled form is displayed
+ * * is [DisplayMode.DisplayAccessCard] the form is hidden and member information is presented with
+ * barcode.
+ *
+ *
+ * @author Sameer Dhakal (Fuzz)
+ */
+class AccessMemberCardViewModel @Inject constructor(
+        private val service: RetrofitMemberDataProvider,
+        private val infoPreferencesManager: MemberInfoPreferencesManager,
+        private val analyticsTracker: AnalyticsTracker
+) : BaseViewModel() {
+
+    sealed class DisplayMode {
+        object DisplayForm : DisplayMode()
+        class DisplayAccessCard(val memberID: String) : DisplayMode()
+        class UpdateForm(val memberID: String, val zipCode: String) : DisplayMode()
+    }
+
+    /**
+     * Emits events for various loading states.
+     */
+    var loadStatus: BehaviorSubject<LoadStatus> = BehaviorSubject.createDefault<LoadStatus>(LoadStatus.None())
+
+    /**
+     * Form field events.
+     */
+    var zipCode = BehaviorSubject.createDefault<String>("")
+    var memberID = BehaviorSubject.createDefault<String>("")
+    var isValid = BehaviorSubject.createDefault<Boolean>(false)
+
+
+    var displayMode: Subject<DisplayMode> = BehaviorSubject.create()
+
+
+    private var members: Subject<List<MemberInfo>> = BehaviorSubject.create()
+    private var selectedMember: Subject<MemberInfo> = BehaviorSubject.create()
+
+    var expiration: Subject<String> = BehaviorSubject.create()
+    var selectedCardHolder: Subject<String> = BehaviorSubject.create()
+    var membership: Subject<String> = BehaviorSubject.create()
+
+    init {
+
+        /**
+         * Validate the form and enable the sign in button.
+         */
+        Observables.combineLatest(
+                zipCode,
+                memberID)
+                .map { (zipCode, memberID) ->
+                    zipCode.isNotEmpty() && memberID.isNotEmpty()
+
+                }
+                .bindTo(isValid)
+                .disposedBy(disposeBag)
+
+        /**
+         * Updated expiration and membership data when cardHolder is switched.
+         */
+        selectedMember
+                .subscribeBy { memberInfo ->
+                    expiration.onNext(memberInfo.expiration.orEmpty())
+                    selectedCardHolder.onNext(memberInfo.cardHolder.orEmpty())
+                    membership.onNext(memberInfo.memberLevel.orEmpty())
+
+                    memberInfo.primaryConstituentID?.let {
+                        displayMode.onNext(DisplayMode.DisplayAccessCard(it))
+                        /**
+                         * Log show_card analytics event.
+                         */
+                        analyticsTracker.reportEvent(EventCategoryName.Member, AnalyticsAction.memberShowCard)
```

Comment:
Would it work ok if we did the analytics event _before_ `displayMode.onNext`? Or would that mess up timing?

Reply:
It does not matter. Could you please mention the reason why should analytics event happen before the `displayMode.onNext` ?

Reply:
It only matters if we expect `.onNext` to trigger a screenview or other event. In that case keeping the sequence of analytics events consistent is helpful.

Reply:
I could've been clearer in my original comment: your code as it stands is perfectly acceptable if there are no other events associated with switching to that mode.

Reply:
Actually, the correct place to trigger analytics event is when `DisplayMode` is `DisplayAccessCard`. Updated.